modules/logging/formattedencoder.go

```diff
@@ -0,0 +1,147 @@
+			}
+		},
+	}
+}
+
+// Provision sets up the encoder.
+func (se *FormattedEncoder) Provision(ctx caddy.Context) error {
+	if se.Template == "" {
+		return fmt.Errorf("missing template for formatted log encoder")
+	}
+	se.Encoder = zapcore.NewJSONEncoder(se.ZapcoreEncoderConfig())
+	return nil
+}
+
+// Clone wraps the underlying encoder's Clone. This is
+// necessary because we implement our own EncodeEntry,
+// and if we simply let the embedded encoder's Clone
+// be promoted, it would return a clone of that, and
+// we'd lose our FormattedEncoder's EncodeEntry.
+func (se FormattedEncoder) Clone() zapcore.Encoder {
+	return FormattedEncoder{
+		Encoder:  se.Encoder.Clone(),
+		Template: se.Template,
+	}
+}
+
+// EncodeEntry partially implements the zapcore.Encoder interface.
+func (se FormattedEncoder) EncodeEntry(ent zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	repl := caddy.NewReplacer()
+	buf, err := se.Encoder.EncodeEntry(ent, fields)
+	if err != nil {
+		return nil, err
+	}
+	appName := strings.SplitN(ent.LoggerName, ".", 2)[0]
+	// set the vals in the replacer
+	err = jsonparser.ObjectEach(buf.Bytes(), visitor(appName, repl))
+	buf.Reset() // the buffer is only used to collect placeholders' values anyway
+	buf.Free()
+	if err != nil {
+		return nil, err
+	}
+
+	out := repl.ReplaceKnown(se.Template, "")
+	// Unescape escaped quotes
+	buf.AppendString(strings.Replace(out, `\"`, `"`, -1))
+	if !strings.HasSuffix(out, "\n") {
+		buf.AppendByte('\n')
+	}
+
+	return buf, err
+}
+
+func visitor(prefix string, repl *caddy.Replacer) func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
+	format := fmt.Sprintf("%s.%%s", prefix)
+	keyFormat := func(itemKey string) string {
+		return fmt.Sprintf(format, itemKey)
+	}
+	return func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
+		switch dataType {
+		case jsonparser.NotExist:
+			panic("unimplemented jsonparser visitor for NotExist")
+		case jsonparser.String:
+			repl.Set(keyFormat(string(key)), string(value))
+		case jsonparser.Number:
+			// see: https://github.com/buger/jsonparser/issues/85
+			if bytes.IndexByte(value, '.') > -1 {
+				v, _ := strconv.ParseFloat(string(value), 64)
+				repl.Set(keyFormat(string(key)), v)
+			} else {
+				v, _ := strconv.ParseUint(string(value), 10, 64)
+				repl.Set(keyFormat(string(key)), v)
+			}
+		case jsonparser.Object:
+			// recurse
+			err := jsonparser.ObjectEach(value, visitor(keyFormat(string(key)), repl))
+			if err != nil {
+				return err
+			}
+		case jsonparser.Array:
+			repl.Set(keyFormat(string(key)), value)
+		case jsonparser.Boolean:
+			v, _ := strconv.ParseBool(string(value))
+			repl.Set(keyFormat(string(key)), v)
+		case jsonparser.Null:
+			panic("unimplemented jsonparser visitor for Null")
+		case jsonparser.Unknown:
+			panic("unimplemented jsonparser visitor for Unknown")
+		default:
+			panic("completely unknown dataType")
+		}
```

Comment:
How should these cases (and `jsonparser.NotExist` above) be handled?

Reply:
Hm, yeah instead of panicking we should probably just return an empty value. With structured logs we can't guarantee that a field will exist, so gracefully treat it as empty.

Reply:
The `panic` isn't meant to stay. It's meant tingle your spidey senses and get you to comment on it for feedback 😛  You're suggesting `repl.Set(keyFormat(string(key)), "")` this, right?